src/analyzeFile.ts

```diff
@@ -0,0 +1,95 @@
+import {
+  Conditions,
+  Diagnostic,
+  DiagnosticError,
+  ValidateContext,
+} from "./types";
+import { startsWithKeyword } from "./utils";
+import { after_date } from "./after_date";
+import { on_branch } from "./on_branch";
+import { when } from "./when";
+
+const conditions: Conditions = {
+  after_date,
+  on_branch,
+  when,
+};
+
+function isDiagnosticError(
+  validation: Diagnostic
+): validation is DiagnosticError {
+  return validation.error === true;
+}
+
+export function analyzeFile(
+  file: string,
+  { keywords, options, packageJson }: ValidateContext
+): DiagnosticError[] {
+  const lines = file
+    .split("\n")
+    .map((line, index): [string, number] => [line, index])
+    .filter(([line]) => startsWithKeyword(line.trim(), keywords));
+
+  return lines
+    .map(([line, lineNumber]): Diagnostic => {
```

Comment:
Maybe same here transform and filter in one traverse of the array...

Reply:
I went for readability over performance. Do you think it will make a big difference in performance? Is it worth it to hand in readability?

Reply:
Same remark here..when we traverse a file with 5000 lines or more or maybe multiple large files will make some differences